File: fl_frontend/src/core/schemas.ts
import z from "zod";

import { CustomObjectIdStringSchema } from "./objectId";

/**
 * The envelope every FastAPI response carries, mirroring the backend's `BaseAPIResponse`.
 *
 * Not in `api.ts`: that module is `server-only`, and every feature `schemas.ts` value-imports this
 * envelope, which would taint them all.
 */
export const BaseAPIResponseSchema = z.object({ acknowledged: z.literal([0, 1]) });
export type BaseAPIResponse = z.infer<typeof BaseAPIResponseSchema>;

/** Every failure's body, mirroring the backend's `FLFailureBody` (`docs/logging/spec.md :: L4`). */
export const FLFailureBodySchema = z.object({ error_code: z.string(), trace_id: z.string() });

/** One refusal a `REQ-VAL-001` names; `kind` is pydantic's error `type`. */
export const FLRefusedFieldSchema = z.object({
  in: z.enum(["body", "query", "path", "header", "cookie"]),
  path: z.array(z.union([z.string(), z.number().int()])),
  kind: z.string(),
});
export type FLRefusedField = z.infer<typeof FLRefusedFieldSchema>;

export const FLRefusedPayloadBodySchema = FLFailureBodySchema.extend({ fields: z.array(FLRefusedFieldSchema) });

/**
 * Which mailbox `POST /identitaet/subjekt` is asked about, folded
 * (`fl_frontend/src/core/emailAddress.ts :: asSignInIdentifier`). No length or alphabet is restated
 * here: `core` may not import `fl_frontend/src/shared/schemas.ts`
 * (`fl_frontend/eslint.config.mjs :: LAYER_BOUNDARY`), and a second copy would drift unwatched.
 */
export const FLSubjektPayloadSchema = z.object({ email: z.string() });
export type FLSubjektPayload = z.infer<typeof FLSubjektPayloadSchema>;

/**
 * One contact seat the mailbox holds on a `saison_teams` row, per seat rather than per person: one
 * junction row seats one person twice where `trainer_ist_zugleich` says so.
 */
export const FLSubjektSitzSchema = z.object({
  saison_id: z.string(),
  team_id: CustomObjectIdStringSchema,
  // The contact-seat sense of the word, never the captaincy's
  // (`fl_backend/app/api/bewerbungen/schemas.py :: FLKontaktRolle`).
  rolle: z.enum(["trainer", "ansprechperson", "stellvertretung"]),
  // The name the club was PLAYED under, never the one it carries now
  // (`docs/backend/spec.md :: I13`).
  team_name: z.string(),
  saison_status: z.enum(["past", "active", "future"]),
});
export type FLSubjektSitz = z.infer<typeof FLSubjektSitzSchema>;

/** One pupil record the mailbox names, the id alone: the caller composed the address it asked about. */
export const FLSubjektSpielerSchema = z.object({ spieler_id: CustomObjectIdStringSchema });
export type FLSubjektSpieler = z.infer<typeof FLSubjektSpielerSchema>;

/** One referee record the mailbox names, carrying the id alone for `FLSubjektSpielerSchema`'s reason. */
export const FLSubjektSchiedsrichterSchema = z.object({ schiedsrichter_id: CustomObjectIdStringSchema });
export type FLSubjektSchiedsrichter = z.infer<typeof FLSubjektSchiedsrichterSchema>;

/**
 * Which confirmed, live league records one mailbox matches. A list under each rather than an
 * optional record: one inbox holds seats at two clubs, and two pupils share an address
 * (`docs/datenschutz.md`).
 */
export const FLSubjektResponseSchema = BaseAPIResponseSchema.extend({
  sitze: z.array(FLSubjektSitzSchema),
  spieler: z.array(FLSubjektSpielerSchema),
  schiedsrichter: z.array(FLSubjektSchiedsrichterSchema),
  // Read beside the lists and never from their being empty: empty lists with this set are a person
  // one confirmation from holding their records, not a mailbox the league holds nothing for
  // (`docs/backend/spec.md :: I374`).
  unbestaetigt: z.boolean(),
});
export type FLSubjektResponse = z.infer<typeof FLSubjektResponseSchema>;

File: fl_backend/app/api/identitaet/schemas.py
from pydantic import BaseModel, ConfigDict

# The wire's spelling of the three seats, imported rather than restated as
# `app/api/kontakte/schemas.py` imports it: a second closed set here would name a seat no other
# endpoint publishes.
from app.api.bewerbungen.schemas import FLKontaktRolle
from app.api.saisons.schemas import FLSaisonStatus
from app.shared.schemas.custom import CustomObjectId
from app.shared.schemas.kontakt import CustomEmail
from app.shared.schemas.responses import BaseAPIResponse


class FLSubjektPayload(BaseModel):
    """Which mailbox is asking, folded to the one spelling the join, the header and the log agree on."""

    model_config = ConfigDict(extra="forbid")

    # In the BODY and on no path or query: an address in a URL reaches the edge's access line, which
    # `docs/logging/spec.md :: L11` keeps a credential out of and an identifier is no better in.
    email: CustomEmail


class FLSubjektSitz(BaseModel):
    """One contact seat the mailbox holds on a season's junction row.

    Per seat and never per person, as `app/api/kontakte/schemas.py :: FLKontaktSitz` is: one row
    seats one person twice where `trainer_ist_zugleich` says so.
    """

    saison_id: str
    team_id: CustomObjectId
    # The contact-seat sense of the word, never the captaincy's, which `docs/glossary.md` holds
    # apart (`app/api/spieler/schemas.py :: FLSpielerRolle`).
    rolle: FLKontaktRolle
    # The junction row's own name, the one the club was PLAYED under (`docs/backend/spec.md :: I13`):
    # a club renamed since is still named here as its seats were held.
    team_name: str
    # The one read this lookup makes of `saisons`. The current-season narrowing is the caller's, and
    # the status is what lets it narrow from this answer rather than from a second round trip.
    saison_status: FLSaisonStatus


class FLSubjektSpieler(BaseModel):
    """One pupil record the mailbox names.

    The id alone: the caller composed the address it asked about, and anything else here hands a
    machine read a fresh copy of somebody's own data for nothing.
    """

    spieler_id: CustomObjectId


class FLSubjektSchiedsrichter(BaseModel):
    """One referee record the mailbox names, carrying the id alone for `FLSubjektSpieler`'s reason."""

    schiedsrichter_id: CustomObjectId


class FLSubjekt(BaseModel):
    """Which confirmed, live records one mailbox matches, as three lists rather than three optional records.

    A list under each because one inbox holds seats at two clubs and two pupils share an address
    (`docs/datenschutz.md :: "Colleagues sharing a school inbox"`).
    """

    sitze: list[FLSubjektSitz]
    spieler: list[FLSubjektSpieler]
    schiedsrichter: list[FLSubjektSchiedsrichter]
    # Its own field and never read off empty lists: set, the mailbox holds records that could grant a
    # panel and none is confirmed, which empty lists alone would report as nothing held
    # (`docs/backend/spec.md :: I374`).
    unbestaetigt: bool


class FLSubjektResponse(FLSubjekt, BaseAPIResponse):
    """Which confirmed, live records one mailbox matches, and whether the address is on the ban list.

    The ban narrows none of the records: whether a barred person signs in is the sign-in gate's to decide.
    """

    # The endpoint's alone and not `FLSubjekt`'s: a person endpoint judging a Funktion reads the
    # records, and the ban is keyed under a secret only this operation needs (`docs/backend/spec.md :: I389`).
    gesperrt: bool
